### DISCUSSION TITLE
Interrupt Service Routines

### DISCUSSION THREAD

**Discussion by 0xF0F0F0F0** [DevSkillsHQ/discusssions#57] [Help]
I am trying to implement an ISR function, but it is not clear to me how the interrupt handler works in the C unit test. 
Since ISRs are dispatched by the hardware based on certain flags, i would imagine that some kind of emulation for this behavior is required when running the tests; I don't see this in the boilerplate code (unless CUnit does it behind the scenes). Could you please clarify?

**Comment by andikovacs** [comment]
Hello @0xF0F0F0F0 

You are right on the real hardware the ISR is triggered by the hardware. In the unit test environment, the unit test itself calls the ISR when certain events are supposed to happen. If for example the UART is supposed to send a character then the TDR register is populated with that character. Then the unit test supposes that after a while that the character transmission was successful and calls the ISR. In parallel with this the unit test asserts that some registers changed their values accordingly. If the send_uart function and the isr are implemented according to the requirements the unit tests shall successfully pass. The unit tests are available and can be used as additional "documentation" to understand the expected behavior.
